networkx/algorithms/dag.py

```diff
@@ -102,6 +102,81 @@ def is_directed_acyclic_graph(G):
     return G.is_directed() and not has_cycle(G)
 
 
+def topological_generations(G):
+    """Stratifies a DAG into generations.
+    
+    A topological generation is node collection in which ancestors of a node in each
+    generation are guaranteed to be in a previous generation, and any descendants of
+    a node are guaranteed to be in a following generation. Nodes are guaranteed to
+    be in the earliest possible generation that they can belong to.
+
+    Parameters
+    ----------
+    G : NetworkX digraph
+        A directed acyclic graph (DAG)
+
+    Returns
+    -------
+    iterable[set]
+        An iterable of sets of nodes.
+
+    Raises
+    ------
+    NetworkXError
+        Generations are defined for directed graphs only. If the graph
+        `G` is undirected, a :exc:`NetworkXError` is raised.
+
+    NetworkXUnfeasible
+        If `G` is not a directed acyclic graph (DAG) no topological generations
+        exist and a :exc:`NetworkXUnfeasible` exception is raised.  This can also
+        be raised if `G` is changed while the returned iterator is being processed
+
+    RuntimeError
+        If `G` is changed while the returned iterator is being processed.
+
+    Examples
+    --------
+    >>> DG = nx.DiGraph([(2, 1), (3, 1)])
+    >>> [sorted(generation) for generation in nx.topological_generations(DG)]
+    [[2, 3], [1]]
+
+    Notes
+    -----
+    The generation in which a node resides can also be determined by taking the
+    max-path-distance from the node to the farthest leaf node.
+
+    See also
+    --------
+    topological_sort
+    """
+    if not G.is_directed():
+        raise nx.NetworkXError("Topological sort not defined on undirected graphs.")
+
+    indegree_map = {v: d for v, d in G.in_degree() if d > 0}
+    zero_indegree = {v for v, d in G.in_degree() if d == 0}
+
+    while zero_indegree:
+        this_generation = zero_indegree
+        zero_indegree = set()
+        for node in this_generation:
+            if node not in G:
+                raise RuntimeError("Graph changed during iteration")
+            for _, child in G.edges(node):
```

Comment:
```suggestion
            for child in G.neighbors(node):
```
The neighbors function is the right way to find the other end of the edges from a given node. I know it wasn't that way in the old code, but it is the better way to do it and it helps performance quite a bit too.

Reply:
I see now your comment that this doesn't work with multigraph edges...  e.g. it doesn't repeat the same node twice, so the degree is not reduced by the right amount. I think we could fix this, but it is not the point of this PR. So leaving it as is should be fine.  Sorry for the noise...